contracts/bridge.sol

```diff
@@ -99,63 +107,117 @@ contract HomeBridge {
+        assembly {
+            hash := mload(add(message, 84))
+        }
+        return hash;
+    }
+
+    /// to be called by authorities to check
+    /// whether they withdraw message should be relayed or whether it
+    /// is too low to cover the cost of calling withdraw and can be ignored
+    function isMessageValueSufficientToCoverRelay(bytes message) public view returns (bool) {
+        return getValueFromMessage(message) > getWithdrawRelayCost();
+    }
+
+    /// an upper bound to the cost of relaying a withdraw by calling HomeBridge.withdraw
+    function getWithdrawRelayCost() public view returns (uint) {
+        return estimatedGasCostOfWithdraw * tx.gasprice;
+    }
+
     /// Used to withdraw money from the contract.
     ///
     /// message contains:
     /// withdrawal recipient (bytes20)
     /// withdrawal value (uint)
     /// foreign transaction hash (bytes32) // to avoid transaction duplication
     ///
-    /// NOTE that anyone can call withdraw provided they have the
-    /// message and required signatures!
+    /// NOTE that anyone can call withdraw provided they have the message and required signatures!
     function withdraw (uint8[] v, bytes32[] r, bytes32[] s, bytes message) public allAuthorities(v, r, s, message) {
         require(message.length == 84);
-        address recipient;
-        uint value;
-        bytes32 hash;
-        // solium-disable-next-line security/no-inline-assembly
-        assembly {
-            // layout of message :: bytes:
-            // offset  0: 32 bytes :: uint (little endian) - message length
-            // offset 32: 20 bytes :: address - recipient address
-            // offset 52: 32 bytes :: uint (little endian) - value
-            // offset 84: 32 bytes :: bytes32 - transaction hash
-
-            // we require above that message length == 84.
-            // bytes 1 to 32 are 0 because message length is stored as little endian.
-            // mload always reads 32 bytes.
-            // so we can and have to start reading recipient at offset 20 instead of 32.
-            // if we were to read at 32 the address would contain part of value and be corrupted.
-            // when reading from offset 20 mload will read 12 zero bytes followed
-            // by the 20 recipient address bytes and correctly convert it into an address.
-            // this saves some storage/gas over the alternative solution
-            // which is padding address to 32 bytes and reading recipient at offset 32.
-            // for more details see discussion in:
-            // https://github.com/paritytech/parity-bridge/issues/61
-            recipient := mload(add(message, 20))
-            value := mload(add(message, 52))
-            hash := mload(add(message, 84))
-        }
+        address recipient = getRecipientFromMessage(message);
+        uint value = getValueFromMessage(message);
+        bytes32 hash = getTransactionHashFromMessage(message);
```

Comment:
These function calls are probably not inlined by solidity. Have you compared the gas cost of this function call on master branch vs this branch?

Reply:
overall this PR increases the gas cost of `HomeBridge.withdraw` by `1634` (`2%`) from `71715` (master: https://travis-ci.org/paritytech/parity-bridge/jobs/318131779#L1221) to `73349` (snd-issue-54-attack-vector: https://travis-ci.org/paritytech/parity-bridge/jobs/326815388#L1279). this increase includes the function calls as well as the added functionality (calculation of cost and transfer to sender). imho this increase is  acceptable.